moveit_ros/planning/kinematics_plugin_loader/src/kinematics_plugin_loader.cpp

```diff
@@ -239,8 +233,9 @@ class KinematicsPluginLoader::KinematicsLoaderImpl
   std::map<std::string, std::vector<std::string> > iksolver_to_tip_links_;  // a map between each ik solver and a vector
                                                                             // of custom-specified tip link(s)
   std::shared_ptr<pluginlib::ClassLoader<kinematics::KinematicsBase> > kinematics_loader_;
-  std::map<const robot_model::JointModelGroup*, std::vector<kinematics::KinematicsBasePtr> > instances_;
+  std::map<const robot_model::JointModelGroup*, kinematics::KinematicsBasePtr> instances_;
```

Comment:
I had the impression that multiple plugins can be loaded for the same JointModelGroup* (for whatever reason). This seems to change that. Is that intentional?

Reply:
Sorry, I only explained the reasoning for this in https://github.com/ros-planning/moveit/pull/150#issuecomment-433106660

The KinematicsPluginLoader caches created kinematics solver (KS) instances. But the cache only maintains **unique** instances - they are not shared. The only purpose of the cache is to not allocate a KS twice in a row - once for validating that it is suitable for JMG [here](https://github.com/ros-planning/moveit/blob/a46f0fb654dee077bcf39c1a2c48f4dffe1068f4/moveit_ros/planning/robot_model_loader/src/robot_model_loader.cpp#L203) and once for actual use [here](https://github.com/ros-planning/moveit/blob/a46f0fb654dee077bcf39c1a2c48f4dffe1068f4/moveit_core/robot_model/src/joint_model_group.cpp#L604). 
Eventually, the JMG owns the created KS instance. KPL only caches it for short term.
As I got the very same impression as @simonschmeisser in the first place, I decided to emphasize the fact that this is only short-term cache by changing the data structure - only the last created KS instance is held per JMG (and passed on to the caller in the 2nd call in a row).
This also resolves a potential issue with dangling shared_ptrs in the cache (which usually should only be weak_ptrs for this reason).

Reply:
Just as a side note: Indeed each JMG owns its unique solver instance. This helps to avoid race conditions when using them from multiple threads in parallel.